scipy/stats/_continuous_distns.py

```diff
@@ -5508,67 +5508,119 @@ def _entropy(self, s):
+            shifted = data - loc
+            return np.sum((1 + np.log(shifted/scale)/shape**2)/shifted)
 
-        if not np.isfinite(data).all():
-            raise ValueError("The data contains non-finite values.")
+        def get_estimate(loc):
+            shape, scale = get_shape_scale(loc)
+            return shape, loc, scale
 
-        floc = float(floc)
-        if floc != 0:
-            # Shifting the data by floc. Don't do the subtraction in-place,
-            # because `data` might be a view of the input array.
-            data = data - floc
-        if np.any(data <= 0):
-            raise FitDataError("lognorm", lower=floc, upper=np.inf)
-        lndata = np.log(data)
-
-        # Three cases to handle:
-        # * shape and scale both free
-        # * shape fixed, scale free
-        # * shape free, scale fixed
+        parameters = _check_fit_input_parameters(self, data, args, kwds)
+        data, fshape, floc, fscale = parameters
+        mn = data.min()
 
-        if fscale is None:
-            # scale is free.
-            scale = np.exp(lndata.mean())
-            if f0 is None:
-                # shape is free.
-                shape = lndata.std()
+        if floc is None:
+            mn_inf = np.nextafter(mn, -np.inf)
+            with np.errstate(over='raise'):
+                try:
+                    std = data.std()
+                except FloatingPointError:
+                    # Handle possible overflows when squaring.
+                    absmx = np.max(np.abs(data))
+                    std = np.std(data/absmx) * absmx
+            # Range of acceptable locations based on sample statistics.
+            lbrack, rbrack = data.mean() - alpha_std*std, mn_inf
+
+            i = 1.
+            delta = mn - mn_inf
+            while dL_dLoc(rbrack) >= 0:
+                i *= 2
+                rbrack = mn - delta*i
+
+            if not (rbrack > lbrack):
+                loc = mn_inf
             else:
-                # shape is fixed.
-                shape = float(f0)
+                # If the derivative is positive at `mn_inf`, and negative a few
+                # `delta` away from it, even though root finding will work, due
+                # to numerical inaccuracy the derivative evaluated at the root
+                # will be much too far away from zero, so that solution is
+                # discarded.
+                has_right_divergence = i > 1
+                b_steps = mn - np.logspace(np.log10(mn-lbrack),
+                                           np.log10(mn-rbrack),
+                                           n_steps)
+
+                derivatives = np.array([dL_dLoc(x) for x in b_steps])
+                sign_changes = np.nonzero(derivatives[1:]*derivatives[:-1] < 0)
+
+                results = [root_scalar(dL_dLoc,
+                                       bracket=(b_steps[idx], b_steps[idx+1]),
+                                       xtol=xtol)
```

Comment:
The default `xtol` (if I am not mistaken, `2e-12` for `brentq`) is usually perfectly fine. It may happen that it is a bit too lenient for this method to beat the default method, in terms of likelihood. Systematically setting it to a lower value may induce an unneeded performance hit. I am not sure that it really matters from an end user point of view, but it could be useful to be able to set it.